### PR TITLE
Rewrote the /ban add sub command

### PR DESCRIPTION
**Pull request by hakusaro**
This is a replacement of the /ban add command to accommodate #1510 and the general poor quality of the ban add command.

There are a couple catches (because I'm replicating the existing functionality and not necessarily fixing the problems):

* If you ban an offline player, you ban *anyone with their character name*. On servers where you can login as a different player, you *don't actually ban that player's character name, just their username*. This is the exact same way the system behaves today. A better solution is a database field for user names and character names and let you ban on both.
* If a player joins the server with an IP address name and that IP address gets banned, they will get banned too. Conversely, if a player with an IP address name joins and gets banned but they just registered, they won't have any known IPs so the ban will "fail" and also "succeed" at the same time. The ban system is great.
----
/ban add now replaces add, addtemp, and addip.
New syntax: /ban add <target> [time] [reason]

Examples:
- /ban add Shank 10d Rewrote the ban system.
- /ban add Ash
- /ban add 127.0.0.1 5d Go work on homework.

Note that if you want to specify a reason and a permaban, you need
to use 0 (zero) as the duration.

Examples:
- /ban add Ash 0 Love ya.
- /ban add Shank 0 Hacking.

### CODE REVIEW

**Comment by mistzzt** [comment]
* A default ban reason may be considered as configurable.
* Can IP ban with no player related moved to a new command like _ipban_?
It would be a good idea to implement IP bans on IP range or CIDR.

**Comment by hakusaro** [comment]
I would rather move those two to #1511 given that this system is on its way out anyway. But that's just me.

**Comment by hakusaro** [comment]
(The point here is more to: Rewrite the add command so that it for sure doesn't match the issues in #1510. I just decided to try to roll account bans into this package because we already have bugs in the current system with them.)

File: TShockAPI/Commands.cs [hunk 1]
```diff
@@ -1285,200 +1286,199 @@ private static void Ban(CommandArgs args)
...
-							else
-								args.Player.SendErrorMessage("Invalid player or account!");
-						}
-						else if (players.Count > 1)
-							TShock.Utils.SendMultipleMatchError(args.Player, players.Select(p => p.Name));
-						else
+						// A ban source is effective root if they are not not real and not rest
+						// Super admins ingame who attempt to run this will have to remove
+						// the immune to ban permission if they want to ban an immune player
+						if (!(args.Player is TSRestPlayer) && !args.Player.RealPlayer)
```
Comment (deadsurgeon42) [inline TShockAPI/Commands.cs]: Isn't it clearer to simply say `if (args.Player is TSServerPlayer)` as `TSServerPlayer` is exclusive to console?

**Comment by hakusaro** [comment]
I'd like to add that while this is somewhat started from #1510, I would like to point out that we do not find similarities enough between Commaster's code and our code to warrant a speedier response. In particular, after analysis, we determined that it was unlikely that the code was copied.

File: TShockAPI/Commands.cs [hunk 2]
```diff
@@ -1285,200 +1286,197 @@ private static void Ban(CommandArgs args)
...
+						// Case: Players & user are invalid, could be IP?
+						// Note: Order matters. If this method is above the online player check,
+						// This enables you to ban an IP even if the player exists in the database as a player.
+						// You'll get two bans for the price of one, in theory, because both IP and user named IP will be banned.
+						// ??? edge cases are weird, but this is going to happen
+						// The only way around this is to either segregate off the IP code or do something else.
+						if (players.Count == 0)
+						{
+							// If the target is a valid IP...
+							string pattern = @"^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$";
```
Comment (Ijwu) [inline TShockAPI/Commands.cs]: Why bother checking if this is a valid IP?

File: TShockAPI/DB/BanManager.cs [hunk 3]
```diff
@@ -46,7 +46,8 @@ public BanManager(IDbConnection db)
 									new SqlColumn("Reason", MySqlDbType.Text),
 									new SqlColumn("BanningUser", MySqlDbType.Text),
 									new SqlColumn("Date", MySqlDbType.Text),
-									new SqlColumn("Expiration", MySqlDbType.Text)
+									new SqlColumn("Expiration", MySqlDbType.Text),
+									new SqlColumn("AccountName", MySqlDbType.Text)
```
Comment (Ijwu) [inline TShockAPI/DB/BanManager.cs]: Have we considered a migration plan for existing databases? Bundling with a program to upgrade SQLite or a script for MySQL may be preferable. I don't expect TShock to do it, but I expect us to address it.

File: TShockAPI/Commands.cs (hunk 2 above)
Comment (hakusaro) [reply TShockAPI/Commands.cs]: Because we shouldn't IP ban any random input and call it a day.

File: TShockAPI/DB/BanManager.cs (hunk 3 above)
Comment (hakusaro) [reply TShockAPI/DB/BanManager.cs]: System automatically adds new database columns to existing databases with no outside action required.

File: TShockAPI/Commands.cs (hunk 2 above)
Comment (Ijwu) [reply TShockAPI/Commands.cs]: That's not what I meant. Why are you checking numerical ranges in IPs and not `[0-255]\.[0-255]\.[0-255]\.[0-255]`?

File: TShockAPI/DB/BanManager.cs (hunk 3 above)
Comment (Ijwu) [reply TShockAPI/DB/BanManager.cs]: 👍 I take it that's a part of the pseudo-ORM we have?

Comment (hakusaro) [reply TShockAPI/DB/BanManager.cs]: @Ijwu yeah the ```EnsureTableStructure``` method actually will append columns to the database schema if not already present. It only works if you're appending on the end of the schema, though, and it doesn't work in reverse iirc.

**Pull request merged**